File: sesa-backend/src/main/java/com/sesa/salud/service/fhir/RdaService.java
/**
 * Servicio principal del ciclo de vida del RDA
 * Generación, persistencia y envío a la plataforma IHCE
 * Resolución 1888 de 2025 — Colombia
 * Autor: Ing. J Sebastian Vargas S
 */
package com.sesa.salud.service.fhir;

import com.sesa.salud.dto.rda.RdaStatusDto;
import com.sesa.salud.dto.RdaRecibidoResumenDto;
import com.sesa.salud.dto.RdaRecibidoDetalleDto;
import com.sesa.salud.entity.Atencion;
import com.sesa.salud.entity.RdaEnvio;
import com.sesa.salud.repository.AtencionRepository;
import com.sesa.salud.repository.RdaEnvioRepository;
import com.sesa.salud.tenant.TenantContextHolder;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class RdaService {

    private final AtencionRepository   atencionRepository;
    private final RdaEnvioRepository   rdaEnvioRepository;
    private final RdaGeneratorService  generatorService;
    private final RdaMinisterioClient  ministerioClient;

    // ═══════════════════════════════════════════════════════════════════════
    //  GENERAR RDA
    // ═══════════════════════════════════════════════════════════════════════

    /**
     * Genera (o regenera) el Bundle FHIR para una atención.
     * Si ya existe un RDA PENDIENTE lo reemplaza; si está ENVIADO/CONFIRMADO, crea uno nuevo.
     */
    @Transactional
    public RdaStatusDto generarRda(Long atencionId, RdaEnvio.TipoRda tipoRda) {
        Atencion atencion = atencionRepository.findById(atencionId)
                .orElseThrow(() -> new RuntimeException("Atención no encontrada: " + atencionId));

        String bundleJson = switch (tipoRda) {
            case CONSULTA_EXTERNA -> generatorService.generarRdaConsultaExterna(atencion);
            case PACIENTE         -> generatorService.generarRdaPaciente(atencion);
            default               -> generatorService.generarRdaConsultaExterna(atencion);
        };

        RdaEnvio rdaEnvio = RdaEnvio.builder()
                .atencionId(atencionId)
                .tipoRda(tipoRda)
                .estadoEnvio(RdaEnvio.EstadoRda.PENDIENTE)
                .bundleJson(bundleJson)
                .fechaGeneracion(Instant.now())
                .tenantSchema(TenantContextHolder.getTenantSchema())
                .build();

        rdaEnvio = rdaEnvioRepository.save(rdaEnvio);
        log.info("RDA generado — id:{} atencion:{} tipo:{}", rdaEnvio.getId(), atencionId, tipoRda);
        return toDto(rdaEnvio);
    }

    /** S11: Genera el RDA de Urgencias para un registro de urgencia. */
    @Transactional
    public RdaStatusDto generarRdaUrgencias(Long urgenciaRegistroId) {
        String bundleJson = generatorService.generarRdaUrgencias(urgenciaRegistroId);
        RdaEnvio rdaEnvio = RdaEnvio.builder()
                .urgenciaRegistroId(urgenciaRegistroId)
                .tipoRda(RdaEnvio.TipoRda.URGENCIAS)
                .estadoEnvio(RdaEnvio.EstadoRda.PENDIENTE)
                .bundleJson(bundleJson)
                .fechaGeneracion(Instant.now())
                .tenantSchema(TenantContextHolder.getTenantSchema())
                .build();
        rdaEnvio = rdaEnvioRepository.save(rdaEnvio);
        log.info("RDA Urgencias generado — id:{} urgenciaRegistroId:{}", rdaEnvio.getId(), urgenciaRegistroId);
        return toDto(rdaEnvio);
    }

    /** S11: Genera el RDA de Hospitalización para un ingreso. */
    @Transactional
    public RdaStatusDto generarRdaHospitalizacion(Long hospitalizacionId) {
        String bundleJson = generatorService.generarRdaHospitalizacion(hospitalizacionId);
        RdaEnvio rdaEnvio = RdaEnvio.builder()
                .hospitalizacionId(hospitalizacionId)
                .tipoRda(RdaEnvio.TipoRda.HOSPITALIZACION)
                .estadoEnvio(RdaEnvio.EstadoRda.PENDIENTE)
                .bundleJson(bundleJson)
                .fechaGeneracion(Instant.now())
                .tenantSchema(TenantContextHolder.getTenantSchema())
                .build();
        rdaEnvio = rdaEnvioRepository.save(rdaEnvio);
        log.info("RDA Hospitalización generado — id:{} hospitalizacionId:{}", rdaEnvio.getId(), hospitalizacionId);
        return toDto(rdaEnvio);
    }

    // ═══════════════════════════════════════════════════════════════════════
    //  ENVIAR AL MINISTERIO
    // ═══════════════════════════════════════════════════════════════════════

    /**
     * Envía el RDA más reciente (PENDIENTE) de una atención al Ministerio.
     */
    @Transactional
    public RdaStatusDto enviarAlMinisterio(Long atencionId, RdaEnvio.TipoRda tipoRda) {
        RdaEnvio rdaEnvio = rdaEnvioRepository
                .findFirstByAtencionIdAndTipoRdaOrderByFechaGeneracionDesc(atencionId, tipoRda)
                .orElseThrow(() -> new RuntimeException(
                        "No hay RDA generado para la atención " + atencionId +
                        ". Genérelo primero."));
        return enviarRdaEnvio(rdaEnvio);
    }

    /**
     * Genera Y envía el RDA en un solo paso (flujo recomendado).
     */
    @Transactional
    public RdaStatusDto generarYEnviar(Long atencionId, RdaEnvio.TipoRda tipoRda) {
        generarRda(atencionId, tipoRda);
        return enviarAlMinisterio(atencionId, tipoRda);
    }

    /** S11: Envía el RDA más reciente de una urgencia al Ministerio. */
    @Transactional
    public RdaStatusDto enviarAlMinisterioUrgencias(Long urgenciaRegistroId) {
        RdaEnvio rdaEnvio = rdaEnvioRepository
                .findFirstByUrgenciaRegistroIdAndTipoRdaOrderByFechaGeneracionDesc(urgenciaRegistroId, RdaEnvio.TipoRda.URGENCIAS)
                .orElseThrow(() -> new RuntimeException(
                        "No hay RDA generado para la urgencia " + urgenciaRegistroId + ". Genérelo primero."));
        return enviarRdaEnvio(rdaEnvio);
    }

    /** S11: Envía el RDA más reciente de una hospitalización al Ministerio. */
    @Transactional
    public RdaStatusDto enviarAlMinisterioHospitalizacion(Long hospitalizacionId) {
        RdaEnvio rdaEnvio = rdaEnvioRepository
                .findFirstByHospitalizacionIdAndTipoRdaOrderByFechaGeneracionDesc(hospitalizacionId, RdaEnvio.TipoRda.HOSPITALIZACION)
                .orElseThrow(() -> new RuntimeException(
                        "No hay RDA generado para la hospitalización " + hospitalizacionId + ". Genérelo primero."));
        return enviarRdaEnvio(rdaEnvio);
    }

    /** S11: Genera y envía RDA de urgencias en un paso. */
    @Transactional
    public RdaStatusDto generarYEnviarUrgencias(Long urgenciaRegistroId) {
        generarRdaUrgencias(urgenciaRegistroId);
        return enviarAlMinisterioUrgencias(urgenciaRegistroId);
    }

    /** S11: Genera y envía RDA de hospitalización en un paso. */
    @Transactional
    public RdaStatusDto generarYEnviarHospitalizacion(Long hospitalizacionId) {
        generarRdaHospitalizacion(hospitalizacionId);
        return enviarAlMinisterioHospitalizacion(hospitalizacionId);
    }

    private RdaStatusDto enviarRdaEnvio(RdaEnvio rdaEnvio) {
        if (rdaEnvio.getEstadoEnvio() == RdaEnvio.EstadoRda.CONFIRMADO) {
            throw new RuntimeException("El RDA ya fue confirmado por el Ministerio.");
        }
        try {
            String idMinisterio = ministerioClient.enviarBundle(rdaEnvio.getBundleJson());
            rdaEnvio.setEstadoEnvio(RdaEnvio.EstadoRda.ENVIADO);
            rdaEnvio.setIdMinisterio(idMinisterio);
            rdaEnvio.setFechaEnvio(Instant.now());
            rdaEnvio.setErrorMensaje(null);
            log.info("RDA enviado al Ministerio — rdaId:{} idMinsalud:{}", rdaEnvio.getId(), idMinisterio);
        } catch (Exception e) {
            rdaEnvio.setEstadoEnvio(RdaEnvio.EstadoRda.ERROR);
            rdaEnvio.setErrorMensaje(e.getMessage());
            rdaEnvio.setReintentos(rdaEnvio.getReintentos() != null ? rdaEnvio.getReintentos() + 1 : 1);
            log.error("Error enviando RDA al Ministerio: {}", e.getMessage());
        }
        rdaEnvio = rdaEnvioRepository.save(rdaEnvio);
        return toDto(rdaEnvio);
    }

    // ═══════════════════════════════════════════════════════════════════════
    //  CONSULTAS
    // ═══════════════════════════════════════════════════════════════════════

    @Transactional(readOnly = true)
    public List<RdaStatusDto> listarPorAtencion(Long atencionId) {
        return rdaEnvioRepository
                .findByAtencionIdOrderByFechaGeneracionDesc(atencionId)
                .stream().map(this::toDto).toList();
    }

    @Transactional(readOnly = true)
    public RdaStatusDto obtenerUltimo(Long atencionId, RdaEnvio.TipoRda tipoRda) {
        return rdaEnvioRepository
                .findFirstByAtencionIdAndTipoRdaOrderByFechaGeneracionDesc(atencionId, tipoRda)
                .map(this::toDto)
                .orElse(null);
    }

    @Transactional(readOnly = true)
    public String obtenerBundleJson(Long rdaId, boolean pretty) {
        RdaEnvio envio = rdaEnvioRepository.findById(rdaId)
                .orElseThrow(() -> new RuntimeException("RDA no encontrado: " + rdaId));
        return envio.getBundleJson();
    }

    /** S17: RDA recibidos de otras IPS (IHCE). Por ahora devuelve lista vacía; integrar con API del Ministerio cuando exista. */
    @Transactional(readOnly = true)
    public List<RdaRecibidoResumenDto> obtenerRdaRecibidosPaciente(Long pacienteId) {
        return List.of();
    }

    /** S17: Detalle de un RDA recibido. Por ahora devuelve null si no hay integración. */
    @Transactional(readOnly = true)
    public RdaRecibidoDetalleDto getRdaRecibidoDetalle(Long id) {
        return null;
    }

    // ═══════════════════════════════════════════════════════════════════════
    //  MAPPER
    // ═══════════════════════════════════════════════════════════════════════

    private RdaStatusDto toDto(RdaEnvio e) {
        return RdaStatusDto.builder()
                .rdaId(e.getId())
                .atencionId(e.getAtencionId())
                .urgenciaRegistroId(e.getUrgenciaRegistroId())
                .hospitalizacionId(e.getHospitalizacionId())
                .tipoRda(e.getTipoRda())
                .estadoEnvio(e.getEstadoEnvio())
                .idMinisterio(e.getIdMinisterio())
                .fechaGeneracion(e.getFechaGeneracion())
                .fechaEnvio(e.getFechaEnvio())
                .fechaConfirmacion(e.getFechaConfirmacion())
                .errorMensaje(e.getErrorMensaje())
                .reintentos(e.getReintentos())
                .build();
    }
}

File: sesa-salud/src/app/core/services/rda.service.ts
/**
 * Servicio Angular para el módulo RDA (Resumen Digital de Atención en Salud)
 * Resolución 1888 de 2025 — IHCE Colombia — HL7 FHIR R4
 * Autor: Ing. J Sebastian Vargas S
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from '../../../environments/environment';

export type TipoRda = 'CONSULTA_EXTERNA' | 'HOSPITALIZACION' | 'URGENCIAS' | 'PACIENTE';
export type EstadoRda = 'PENDIENTE' | 'ENVIADO' | 'CONFIRMADO' | 'ERROR';

export interface RdaStatus {
  rdaId: number;
  atencionId: number;
  /** S11: ID del registro de urgencia cuando tipo es URGENCIAS */
  urgenciaRegistroId?: number;
  /** S11: ID de hospitalización cuando tipo es HOSPITALIZACION */
  hospitalizacionId?: number;
  tipoRda: TipoRda;
  estadoEnvio: EstadoRda;
  idMinisterio: string | null;
  fechaGeneracion: string;
  fechaEnvio: string | null;
  fechaConfirmacion: string | null;
  errorMensaje: string | null;
  reintentos: number;
}

/** S17: Resumen de RDA recibido de otra IPS (IHCE). */
export interface RdaRecibidoResumen {
  id: number;
  pacienteId: number;
  idMinisterio: string | null;
  tipoRda: string;
  fechaAtencion: string | null;
  institucionOrigen: string | null;
  fetchedAt: string | null;
}

/** S17: Detalle de RDA recibido. */
export interface RdaRecibidoDetalle {
  id: number;
  pacienteId: number;
  idMinisterio: string | null;
  tipoRda: string;
  fechaAtencion: string | null;
  institucionOrigen: string | null;
  resumenLegible: string | null;
  fetchedAt: string | null;
}

@Injectable({ providedIn: 'root' })
export class RdaService {

  private readonly base = `${environment.apiUrl}/rda`;

  constructor(private http: HttpClient) {}

  /** Genera el Bundle FHIR para la atención */
  generar(atencionId: number, tipoRda: TipoRda = 'CONSULTA_EXTERNA'): Observable<RdaStatus> {
    const params = new HttpParams().set('tipoRda', tipoRda);
    return this.http.post<RdaStatus>(
      `${this.base}/generar/${atencionId}`, null, { params });
  }

  /** Envía el Bundle al servidor FHIR del Ministerio */
  enviarAlMinisterio(atencionId: number, tipoRda: TipoRda = 'CONSULTA_EXTERNA'): Observable<RdaStatus> {
    const params = new HttpParams().set('tipoRda', tipoRda);
    return this.http.post<RdaStatus>(
      `${this.base}/enviar/${atencionId}`, null, { params });
  }

  /** Genera Y envía en un solo paso */
  generarYEnviar(atencionId: number, tipoRda: TipoRda = 'CONSULTA_EXTERNA'): Observable<RdaStatus> {
    const params = new HttpParams().set('tipoRda', tipoRda);
    return this.http.post<RdaStatus>(
      `${this.base}/generar-y-enviar/${atencionId}`, null, { params });
  }

  /** S11: Genera el RDA de Urgencias para un registro de urgencia */
  generarRdaUrgencia(urgenciaRegistroId: number): Observable<RdaStatus> {
    return this.http.post<RdaStatus>(
      `${this.base}/generar/urgencia/${urgenciaRegistroId}`, null);
  }

  /** S11: Genera y envía RDA de Urgencias en un paso */
  generarYEnviarUrgencia(urgenciaRegistroId: number): Observable<RdaStatus> {
    return this.http.post<RdaStatus>(
      `${this.base}/generar-y-enviar/urgencia/${urgenciaRegistroId}`, null);
  }

  /** S11: Genera el RDA de Hospitalización para un ingreso */
  generarRdaHospitalizacion(hospitalizacionId: number): Observable<RdaStatus> {
    return this.http.post<RdaStatus>(
      `${this.base}/generar/hospitalizacion/${hospitalizacionId}`, null);
  }

  /** S11: Genera y envía RDA de Hospitalización en un paso */
  generarYEnviarHospitalizacion(hospitalizacionId: number): Observable<RdaStatus> {
    return this.http.post<RdaStatus>(
      `${this.base}/generar-y-enviar/hospitalizacion/${hospitalizacionId}`, null);
  }

  /** Lista todos los envíos de una atención */
  listarPorAtencion(atencionId: number): Observable<RdaStatus[]> {
    return this.http.get<RdaStatus[]>(`${this.base}/estado/${atencionId}`);
  }

  /** Obtiene el último RDA de una atención por tipo */
  obtenerUltimo(atencionId: number, tipoRda: TipoRda = 'CONSULTA_EXTERNA'): Observable<RdaStatus> {
    const params = new HttpParams().set('tipoRda', tipoRda);
    return this.http.get<RdaStatus>(`${this.base}/estado/${atencionId}/ultimo`, { params });
  }

  /** Descarga el Bundle FHIR en formato JSON */
  descargarBundle(rdaId: number): Observable<Blob> {
    return this.http.get(`${this.base}/bundle/${rdaId}`, {
      responseType: 'blob',
      headers: { Accept: 'application/fhir+json' }
    });
  }

  /** Info normativa del módulo */
  infoNormativo(): Observable<Record<string, string>> {
    return this.http.get<Record<string, string>>(`${this.base}/info`);
  }

  /** S17: RDA recibidos de otras IPS (IHCE). Lista por paciente. */
  getRdaRecibidosPaciente(pacienteId: number): Observable<RdaRecibidoResumen[]> {
    return this.http.get<RdaRecibidoResumen[]>(`${this.base}/recibidos/paciente/${pacienteId}`);
  }

  /** S17: Detalle de un RDA recibido. */
  getRdaRecibidoDetalle(id: number): Observable<RdaRecibidoDetalle | null> {
    return this.http.get<RdaRecibidoDetalle>(`${this.base}/recibidos/${id}`).pipe(
      catchError(() => of(null))
    );
  }

  /** Helpers UI */
  estadoLabel(estado: EstadoRda): string {
    const labels: Record<EstadoRda, string> = {
      PENDIENTE:   'Pendiente',
      ENVIADO:     'Enviado',
      CONFIRMADO:  'Confirmado',
      ERROR:       'Error'
    };
    return labels[estado] ?? estado;
  }

  estadoColor(estado: EstadoRda): string {
    const colors: Record<EstadoRda, string> = {
      PENDIENTE:  'warning',
      ENVIADO:    'info',
      CONFIRMADO: 'success',
      ERROR:      'danger'
    };
    return colors[estado] ?? 'secondary';
  }
}
